### PR TITLE
[Workflow] Group options into a single workflow step options dataclass

### PR DESCRIPTION
**Pull request by suquark**
## Why are these changes needed?

Currently there are too much arguments & duplications in the workflow code. I group them into a single dataclass. This implementation is isolated from the workflow checkpoint API PR. I want to merge this PR before the checkpoint PR so the reviewers won't feel confused.

This PR also fixes some type annotation errors. It also enforces using named arguments for long argument list, which benefits code changes later.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/workflow/common.py [hunk 1]
```diff
@@ -114,36 +114,84 @@ def calculate_identifier(obj: Any) -> str:
...
     name: str
+    # Workflow step options provided by users.
+    step_options: WorkflowStepRuntimeOptions
     # meta data to store
     user_metadata: Dict[str, Any]
 
     def to_metadata(self) -> Dict[str, Any]:
         f = self.func_body
+        # TODO(suquark): "name" here is different from the "name" field.
+        # Maybe rename it to avoid confusion.
```
Comment (iycheng) [inline python/ray/workflow/common.py:188]: Maybe add more comments about what's the name here?

File: python/ray/workflow/step_executor.py [hunk 2]
```diff
@@ -306,40 +310,44 @@ def _wrap_run(func: Callable, step_type: StepType, step_id: "StepID",
 
 
 @ray.remote(num_returns=2)
-def _workflow_step_executor(step_type: StepType, func: Callable,
-                            context: workflow_context.WorkflowStepContext,
+def _workflow_step_executor(func: Callable, context: "WorkflowStepContext",
                             step_id: "StepID",
                             baked_inputs: "_BakedWorkflowInputs",
-                            catch_exceptions: bool, max_retries: int) -> Any:
+                            options: "WorkflowStepRuntimeOptions") -> Any:
```
Comment (iycheng) [inline python/ray/workflow/step_executor.py:316]: options => runtime_options?

**Comment by iycheng** [comment]
Overall it looks nice! Some comments, most are about naming and test failure.

Really thanks for the update!

**Comment by suquark** [comment]
I also splits "metadata" tests because it sometimes timeout.

**Comment by suquark** [comment]
@iycheng Thank you for comments! The PR is ready again

**Review by fishbone — APPROVED** [review]
Thanks!

**Pull request merged**